doc/arch/adr-001-identity-and-authorization.md

```diff
@@ -0,0 +1,134 @@
+
+## Consequences
+
+- Customizations to the login pages will be relatively straightforward.
+- Customizations to the authentication process (e.g. adding social logins or MFA) will be relatively more complex.
+- We need to come up with a deployment plan that accounts for the provider we've chosen. Which is included below.
+
+## Deployment / Migration Plan
+
+1. Begin development on Django API within staging environment. We could either place the Django app at a subdomain (e.g. api.staging.echolocator.org) or path of the existing site (e.g. staging.echolocator.org/api). We've done both in the past but in this case I'm leaning toward a path because that would avoid cross-origin issues and give us access to Django templates if we want them. It would also put the Django admin in a logical place (staging.echolocator.org/admin as opposed to api.staging.echolocator.org/admin, which seems more confusing). This will include setting up CI deployment to staging as well.
+2. Rework staging signup and authentication flow to target the staging Django server. This may cause login to the staging site to become unusable via normal means for a period of time, so we should make sure to keep anonymous authentication available so that BHA can continue to use the site for demo purposes if they need to.
+3. Write user migration scripts.
+4. Thoroughly test user migration scripts against staging. In this case "migration" is actually just copying, so we can do this non-destructively as many times as we want to be certain that everything works as expected. We should also test migrating the production user database into staging, as that has more users and is more likely to encounter edge cases.
+5. Create an admin-configurable flag to prevent user login or signup temporarily. The primary risk during this deployment process will be data loss if a user changes their data between when we start the migrating and when we finish it, so we should prevent users from logging in while the migration is in process.
+6. Develop and test a method to tell all users to reset their password. This could be accomplished via a script and whatever email service we choose to use, or through BHA infrastructure if they have something like MailChimp.
+
+Once the above are all complete, the deployment plan should look something like this:
+1. Coordinate with BHA on a deployment window
+2. Queue up "reset your password" email blast
+3. At start of window, deploy to production with login/signup flag set to false.
+4. Run user migration scripts
+5. Switch login/signup flag to true
+6. Send "reset your password" email blast
```

Comment:
If we can customize the email per user with a script, we can take advantage of Django's signing support to generate links that will take people directly to a "change your password" site - or maybe even just log them in if we're going passwordless.

- Django docs (they're a bit low-level): https://docs.djangoproject.com/en/4.0/topics/signing/
- A blog post about how to do this for unsubscribe links (which is what I've used it for in the past) http://www.grokcode.com/819/one-click-unsubscribes-for-django-apps/

Reply:
Cool! I've never used Django signing before, that looks really neat.